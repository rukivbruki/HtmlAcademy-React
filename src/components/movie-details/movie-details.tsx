// Libs
import * as React from 'react';
import {connect} from "react-redux";
import {Route, Switch, Link, withRouter} from "react-router-dom";
// Utils
import {FilmsType, FilmType, HistoryType, MatchType} from "../../types";
import Paths from "../../router/paths";
import {Operation as FavoritesOperation} from "../../reducer/favorites/favorites.js";
import {getFilms} from "../../reducer/films/selectors.js";
import withActiveCard from "../../hocs/with-active-card/with-active-card";
import {getIsAuth} from "../../reducer/user/selectors";
import {getFavorites} from "../../reducer/favorites/selectors";
// Components
import MovieList from "../movie-list/movie-list";
import Tabs from "../tabs/tabs";
import MyListButton from "../my-list-button/my-list-button";
import Avatar from "../avatar/avatar";
import Logo from '../logo/logo';

const MovieListWithActiveCard = withActiveCard(MovieList);

interface Props {
  isAuth: boolean;
  history?: HistoryType;
  match?: MatchType;
  cardData: FilmType;
  films: FilmsType;
  onOpenCard: () => void;
  filteredFilms: FilmsType;
  setActivePlayer: () => void;
  updateFavorite: (id: number, isFavorite: 0 | 1, loadFavorites: () => void) => void;
  favorites?: FilmsType;
  loadFavorites: () => void;
}

interface MovieDetailsData {
  name: string;
  genre: string;
  posterImage: string;
  released: number;
  backgroundColor: string;
  backgroundImage: string;
}

const defaultMovieDetailsData: MovieDetailsData = {
  name: ``,
  genre: ``,
  posterImage: ``,
  released: 0,
  backgroundColor: ``,
  backgroundImage: ``,
};

const MovieDetails = ({
  isAuth,
  match,
  history,
  films = [],
  onOpenCard,
  filteredFilms,
  updateFavorite,
  favorites = [],
  loadFavorites,
}: Props) => {
  const {
    path,
    url,
    params: {id}
  } = match;
  const data: FilmType | MovieDetailsData = films.find(({id: movieId}) => movieId.toString() === id) || defaultMovieDetailsData;
  const {
    name,
    genre,
    posterImage,
    released,
    backgroundColor,
    backgroundImage,
  } = data;
  const handlePlayButtonClick = () => {
    history.push(`/player/${id}`);
  };

  const renderAddReview = () => {
    if (isAuth) {
      return (
        <Link to={`${url}/review`} className="btn movie-card__button">
          Add review
        </Link>
      );
    }

    return null;
  };

  const isFavorite = !!favorites.find((film) => film.id === parseInt(id, 10));

  const handleOnClickMyList = () => {
    if (!isAuth) {
      history.push(Paths.LOGIN);
      return;
    }

    updateFavorite(id, isFavorite ? 0 : 1, loadFavorites);
  };

  const filterFilmsByOpenedFilmGenre = (filmsList, openedFilmGenre) => filmsList.filter(({genre: filmGenre}) => filmGenre === openedFilmGenre);
  const deleteCurrentFilm = (filmsList, currentFilmId) => (
    filmsList.filter(({id: filmId}) => currentFilmId !== filmId.toString())
  );

  const getPreparedFilms = (filmsList, filmId, filmGenre) => {
    const result = filterFilmsByOpenedFilmGenre(deleteCurrentFilm(filmsList, filmId), filmGenre);

    return result.slice(0, 4);
  };

  return (
    <>
    <section
      className="movie-card movie-card--full"
      style={{
        backgroundColor
      }}
    >
      <div className="movie-card__hero">
        <div className="movie-card__bg">
          <img src={backgroundImage} alt={name} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <header className="page-header movie-card__head">
          <Logo />

          <div className="user-block">
            <Avatar />
          </div>
        </header>

        <div className="movie-card__wrap">
          <div className="movie-card__desc">
            <h2 className="movie-card__title">{name}</h2>
            <p className="movie-card__meta">
              <span className="movie-card__genre">{genre}</span>
              <span className="movie-card__year">{released}</span>
            </p>

            <div className="movie-card__buttons">
              <button
                className="btn btn--play movie-card__button"
                type="button"
                onClick={handlePlayButtonClick}
              >
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  fill="#d9cd8d"
                >
                  <path d="M8 5v14l11-7z" />
                  <path d="M0 0h24v24H0z" fill="none" />
                </svg>
                <span>Play</span>
              </button>
              <MyListButton
                isFavorite={isFavorite}
                isAuth={isAuth}
                onClick={handleOnClickMyList}
              />
              {renderAddReview()}
            </div>
          </div>
        </div>
      </div>

      <div className="movie-card__wrap movie-card__translate-top">
        <div className="movie-card__info">
          <div className="movie-card__poster movie-card__poster--big">
            <img
              src={posterImage}
              alt={`${name} poster`}
              width="218"
              height="327"
            />
          </div>

          <div className="movie-card__desc">
            <nav className="movie-nav movie-card__nav">
              <ul className="movie-nav__list">
                <li className="movie-nav__item">
                  <Link to={`${url}`} className="movie-nav__link">
                    Overview
                  </Link>
                </li>
                <li className="movie-nav__item">
                  <Link to={`${url}/details`} className="movie-nav__link">
                    Details
                  </Link>
                </li>
                <li className="movie-nav__item">
                  <Link to={`${url}/reviews`} className="movie-nav__link">
                    Reviews
                  </Link>
                </li>
              </ul>
            </nav>

            <div>
              <Switch>
                <Route exact path={`${path}`}>
                  <Tabs {...data} />
                </Route>
                <Route path={Paths.FILMS_$ID_DETAILS}>
                  <Tabs {...data} tab='details'/>
                </Route>
                <Route path={Paths.FILMS_$ID_REVIEWS}>
                  <Tabs {...data} tab='reviews'/>
                </Route>
              </Switch>
            </div>
          </div>
        </div>
      </div>
    </section>

    <div className="page-content">
      <section className="catalog catalog--like-this">
        <h2 className="catalog__title">More like this</h2>
        <MovieListWithActiveCard
          onOpenCard={onOpenCard}
          filteredFilms={getPreparedFilms(filteredFilms, id, genre)}
        />
      </section>
    </div>
  </>
  );
};

const mapStateToProps = (state) => ({
  films: getFilms(state),
  isAuth: getIsAuth(state),
  favorites: getFavorites(state)
});

const mapDispatchToProps = (dispatch) => ({
  loadFavorites() {
    dispatch(FavoritesOperation.loadFavorites());
  },
  updateFavorite: (id, status, cb) => {
    dispatch(FavoritesOperation.setFavoriteStatus(id, status, cb));
  }
});

export {MovieDetails};
export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(React.memo(MovieDetails)));
