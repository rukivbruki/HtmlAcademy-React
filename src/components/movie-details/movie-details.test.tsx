import * as React from "react";
import * as renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {MemoryRouter} from 'react-router';
import NameSpace from "../../reducer/name-space.js";
import {MovieDetails} from './movie-details';

const mockStore = configureStore([]);

const noop = () => void 0;

const cardData = {
  backgroundColor: `red`,
  backgroundImage: `img`,
  description: `desc`,
  director: `dir`,
  genre: `genre`,
  id: 1,
  isFavorite: false,
  name: `film`,
  posterImage: `img`,
  previewImage: `img`,
  previewVideoLink: `video`,
  rating: 1,
  released: 1,
  runTime: 1,
  scoresCount: 1,
  starring: [`test`],
  videoLink: `video`
};

const films = [cardData];

const user = {
  id: 1,
  email: `test@email.com`,
  name: `test user`,
  avatarUrl: `avatar`
};

const props = {
  isAuth: true,
  cardData,
  films,
  onOpenCard: noop,
  filteredFilms: films,
  setActivePlayer: noop,
  updateFavorite: noop,
  loadFavorites: noop,
  match: {
    isExact: false,
    params: {
      id: 1
    },
    path: ``,
    url: ``
  }
};

it(`MovieDetails renders correctly`, () => {
  const store = mockStore({
    [NameSpace.GENRES]: {
      genres: [`genre1`, `genre2`],
      genreFilter: `All genres`
    },
    [NameSpace.FILMS]: {
      films
    },
    [NameSpace.USER]: {
      user
    }
  });

  const tree = renderer
    .create(
        <MemoryRouter initialEntries={[`/test`]} >
          <Provider store={store}>
            <MovieDetails {...props}/>
          </Provider>
        </MemoryRouter>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
