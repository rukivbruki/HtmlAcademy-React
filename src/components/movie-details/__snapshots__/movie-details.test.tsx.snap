// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails renders correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": "",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt=""
          src=""
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link logo__link--light"
            href="/"
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <a
            className="sign-in__link"
            href="/login"
            onClick={[Function]}
          >
            Sign In
          </a>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              
            </span>
            <span
              className="movie-card__year"
            >
              0
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="#d9cd8d"
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5v14l11-7z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                onClick={[Function]}
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt=" poster"
            height="327"
            src=""
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/"
                  onClick={[Function]}
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/details"
                  onClick={[Function]}
                >
                  Details
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  href="/reviews"
                  onClick={[Function]}
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div>
            <div
              className="movie-rating"
            >
              <div
                className="movie-rating__score"
              />
              <p
                className="movie-rating__meta"
              >
                <span
                  className="movie-rating__level"
                >
                  -
                </span>
                <span
                  className="movie-rating__count"
                >
                  undefined ratings
                </span>
              </p>
            </div>
            <div
              className="movie-card__text"
            >
              <p />
              <p
                className="movie-card__director"
              >
                <strong>
                  Director: undefined
                </strong>
              </p>
              <p
                className="movie-card__starring"
              >
                <strong>
                  
                </strong>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      />
    </section>
  </div>,
]
`;
